pkg/controllers/multicluster/globalservice_controller.go

```diff
@@ -0,0 +1,239 @@
+/*
+ Copyright 2021 The Hybridnet Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package multicluster
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	multiclusterv1 "github.com/alibaba/hybridnet/pkg/apis/multicluster/v1"
+
+	"github.com/alibaba/hybridnet/pkg/utils"
+
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1beta1 "k8s.io/api/discovery/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/alibaba/hybridnet/pkg/constants"
+	"github.com/alibaba/hybridnet/pkg/controllers/concurrency"
+)
+
+const ControllerGlobalService = "GlobalService"
+
+// GlobalServiceReconciler reconciles a global Service object
+type GlobalServiceReconciler struct {
+	context.Context
+	client.Client
+
+	concurrency.ControllerConcurrency
+}
+
+func (r *GlobalServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+	log := ctrllog.FromContext(ctx)
+
+	defer func() {
+		if err != nil {
+			log.Error(err, "reconciliation fails")
+		}
+	}()
+
+	var service = &corev1.Service{}
+	if err = r.Get(ctx, req.NamespacedName, service); err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, wrapError("unable to get service", err)
+	} else if apierrors.IsNotFound(err) {
+		// endpoints will be cleaned by owner reference
+		return ctrl.Result{}, nil
+	}
+
+	if !isGlobalService(service) {
+		return ctrl.Result{}, wrapError("unable to unbind remote endpoint slices",
+			r.unbindRemoteEndpointSlices(ctx, service))
+	}
+
+	return ctrl.Result{}, wrapError("unable to bind remote endpoint slices",
+		r.bindRemoteEndpointSlices(ctx, service))
+}
+
+func (r *GlobalServiceReconciler) unbindRemoteEndpointSlices(ctx context.Context, svc *corev1.Service) error {
+	// delete all the endpoint slices created by hybridnet
+	epsList := &discoveryv1beta1.EndpointSliceList{}
+	if err := r.List(ctx, epsList, client.MatchingLabels{
```

Comment:
nit: try `r.DeleteAllOf()` instead of listing and deleting

Reply:
done